diffkemp/simpll/PatternComparator.h

```diff
@@ -0,0 +1,132 @@
+    InstructionSet AllInstMatches;
+    /// Mapping for matched module instructions. Instructions from the left
+    /// module are used as keys, while instructions from the right module
+    /// are used as values. Contains the result of a single pattern match.
+    mutable InstructionMap InstMappings;
+
+    /// Tries to match a difference pattern starting with the given instruction
+    /// pair. Returns true if a valid match is found.
+    bool matchPattern(const Instruction *InstL, const Instruction *InstR);
+
+    /// Tries to match a pair of values to a value pattern. Returns true if a
+    /// valid match is found.
+    bool matchValues(const Value *L, const Value *R);
+
+  private:
+    /// Pair of corresponding instruction pattern function comparators.
+    using InstPatternComparatorPair =
+            std::pair<std::unique_ptr<InstPatternComparator>,
+                      std::unique_ptr<InstPatternComparator>>;
+    /// Pair of corresponding value pattern function comparators.
+    using ValuePatternComparatorPair =
+            std::pair<std::unique_ptr<ValuePatternComparator>,
+                      std::unique_ptr<ValuePatternComparator>>;
+
+    /// Parent DifferentialFunctionComparator the compares the module functions.
+    const DifferentialFunctionComparator *DiffFunctionComp;
+    /// Instruction pattern function comparators associated with the current set
+    /// of patterns and the currently compared functions.
+    DenseMap<const InstPattern *, InstPatternComparatorPair> InstPatternComps;
+    /// Value pattern function comparators associated with the current set of
+    /// patterns and the currently compared functions.
+    DenseMap<const ValuePattern *, ValuePatternComparatorPair>
+            ValuePatternComps;
+
+    /// Tries to match one of the loaded instruction patterns. Returns true
+    /// if a valid match is found.
+    bool matchInstPattern(const Instruction *InstL, const Instruction *InstR);
+
+    /// Tries to match one of the loaded value patterns. Returns true if
+    /// a valid match is found.
+    bool matchValuePattern(const Instruction *InstL, const Instruction *InstR);
+
+    /// Tries to match a load instruction to the start of the given value
+    /// pattern.
+    bool matchLoadInst(const LoadInst *Load,
+                       const ValuePattern *Pat,
+                       bool IsLeftSide);
+
+    /// Check whether the input mapping generated by the given pattern function
+    /// comparator pair is valid even when both compared modules are analysed at
+    /// once.
+    bool inputMappingValid(const InstPattern *Pat,
+                           const InstPatternComparatorPair *PatternComps);
+
+    /// Create the resulting instruction mapping and add all matched
+    /// instructions into the combined instruction set.
+    void processPatternMatch(const InstPattern *Pat,
+                             const InstPatternComparatorPair *PatternComps);
```

Comment:
I have considered that. However, since the `InstPatternComparatorPair` type is necessary either way (the type is used to define maps of loaded patterns), I think the original version is more readable. You are right that one of the comparators used is, in essense, `this`. Staticity could technically be used here, but again, I think doing so would make the code less readable.